README.md
# PostCSS Customprop Validate [![Build](https://img.shields.io/github/workflow/status/gojek/postcss-customprop-validate/Test)](https://github.com/gojek/postcss-customprop-validate/actions/workflows/test.yml)

[PostCSS] plugin to validate [fallback values] of CSS custom properties.

Validating the fallback values of CSS custom properties manually can be error prone and challening, specially if there are many.

This plugin validates the CSS and returns custom properties with incorrect fallback values. It can also modify the source CSS file with the updated values.

## Usage

```js
const postcss = require("postcss");
const plugin = require("postcss-customprop-validate");
const { readFileSync } = require("fs");

const cssPath = "./sample.css";
const css = readFileSync(cssPath, "utf-8");
/*
{
  font-size: 1rem;
  color: var(--color-red, #fa0000);
}
*/

const properties = {
  "--color-red": "#ff0000",
};

const callback = (err, data) => {
  if (err) {
    console.error(err);
    return;
  }
  console.log(data);
};

postcss([
  plugin({
    properties,
    write: false,
    callback,
  }),
])
  .process(css, { from: cssPath })
  .then((result) => {
    console.log(result.wrongProps);
    /*
     [
       {
         path: "./sample.js",
         line: 3,
         key: "--color-red",
         current: "#fa0000",
         expected: "#ff0000"
       }
     ]
    */
  });
```

See [PostCSS] docs for examples for your environment.

## Options

The plugin accepts an object containing 3 properties:

### **`properties`**

Required: `true`

An object containg key and expected fallback values of custom properties. The plugin uses it as the source of truth to validate the CSS.

Example:

```json
{
  "--border-radius": "4px",
  "--color-red": "#ff0000",
  "--danger-border":
    "var(--border-radius, 4px) solid var(--color-red, #ff0000)",
};
```

### **`write`**

Default: `false`

Modifies the source CSS file with the correct fallback values.

#### Formatting

The plugin does not preserves the code formatting. If you use a code formatter like [prettier], you may need to re-run it after the plugin modifies the source file.

### **`callback`**

Default: `() => {}`

Callback to handle the output of the plugin. The plugin invokes the callback after it processes the CSS of each file. The callback accepts 2 arguments. The first argument is an error object. The second argument is an array of objects. Each object represents an incorrect fallback value having the following shape:

- `path` - absolute path of the CSS file
- `line` - line number containing the custom property
- `key` - key of the custom property
- `current` - current fallback value
- `expected` - expected fallback value

The second argument is also available in the plugin's result as `result.wrongProps`

## Example

A sample CSS:

```css
.foo {
  border: var(
    --danger-border,
    var(--border-radius, 5px) solid var(--color-red, #fa0000)
  );
}

.bar {
  --danger-text: var(--color-red, red);
  color: var(--danger-text);
}
```

The resultant CSS that the plugin produces:

```diff
.foo {
  border: var(
    --danger-border,
-    var(--border-radius, 5px) solid var(--color-red, #fa0000)
+    var(--border-radius, 4px) solid var(--color-red, #ff0000)
  );
}

.bar {
-  --danger-text: var(--color-red, red);
+  --danger-text: var(--color-red, #ff0000);
  color: var(--danger-text);
}
```

Custom properties with incorrect fallback values:

```js
[
  {
    path: undefined,
    line: 2,
    key: "--danger-border",
    current: "var(--border-radius, 5px) solid var(--color-red, #fa0000)",
    expected: "var(--border-radius, 4px) solid var(--color-red, #ff0000)",
  },
  {
    path: undefined,
    line: 9,
    key: "--color-red",
    current: "red",
    expected: "#ff0000",
  },
];
```

[postcss]: https://github.com/postcss/postcss
[fallback values]: https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties#custom_property_fallback_values
[prettier]: https://github.com/prettier/prettier
[options]: #options
[response]: #response
[official docs]: https://github.com/postcss/postcss#usage


CHANGELOG.md
# Change Log

## 1.0.0 (2021-05-24)

### Added

Inital Version


index.test.js
// const postcss = require('postcss')

// const plugin = require('./')

// async function run (input, output, opts = { }) {
//   let result = await postcss([plugin(opts)]).process(input, { from: undefined })
//   expect(result.css).toEqual(output)
//   expect(result.warnings()).toHaveLength(0)
// }

describe("postcss-cusomprop-validate", () => {
  test.todo("first test")
})

/* Write tests here

it('does something', async () => {
  await run('a{ }', 'a{ }', { })
})

*/


__tests__/index.test.js
const postcss = require("postcss");

jest.mock("fs", () => {
  return {
    writeFile: jest
      .fn()
      .mockImplementation((file, path, cb) => cb("foo"))
      .mockName("nodeWriteFile"),
  };
});

const plugin = require("../index");

async function run(input, opts = {}) {
  return await postcss([plugin(opts)]).process(input, {
    from: undefined,
  });
}

describe("postcss-validate-customprop", () => {
  afterAll(() => {
    jest.resetModules();
  });

  test("validates fallback values of custom properties", async () => {
    const properties = {
      "--border-radius": "4px",
      "--color-red": "#FF0000",
      "--custom-border":
        "var(--border-radius, 4px) solid var(--color-red, #FF0000)",
    };

    const css = `.danger {
      --danger-radius: var(--border-radius, 4px);
      border-radius: var(--danger-radius);
    }
    .success {
      border: var(--custom-border, var(--border-radius, 5px) solid var(--color-red, #FA0000));
    }
    .info {
      color: var(--color-red, red);
    }`;

    const wrongProps = [
      {
        path: undefined,
        line: 6,
        key: "--custom-border",
        current: "var(--border-radius, 5px) solid var(--color-red, #FA0000)",
        expected: "var(--border-radius, 4px) solid var(--color-red, #FF0000)",
      },
      {
        path: undefined,
        line: 9,
        key: "--color-red",
        current: "red",
        expected: "#FF0000",
      },
    ];

    const callback = jest.fn();

    const result = await run(css, { properties, callback });
    expect(result.wrongProps).toEqual(wrongProps);
    expect(callback).toHaveBeenCalledWith(null, wrongProps);

    const cssWithoutCustomProps = `.danger {
      color: hsla(292, 56%, 33%, 1);
      font-size: 1.2rem;
    }`;

    const resultWithoutCustomProps = await run(cssWithoutCustomProps, {
      properties,
    });
    expect(resultWithoutCustomProps.wrongProps).toEqual([]);

    const cssWithUnknownCustomProp = `.success {
      color: var(--color-green, #00FFAA);
      border-radius: var(--border-radius, 4px)
    }`;

    const resultWithUndefinedCustomProps = await run(cssWithUnknownCustomProp, {
      properties,
    });
    expect(resultWithUndefinedCustomProps.wrongProps).toEqual([]);
  });

  test("throws error on missing options", async () => {
    const css = `.danger {
      border-radius: var(--border-radius, 5px);
    }`;

    expect(() =>
      postcss([plugin({})]).process(css, {
        from: undefined,
      })
    ).toThrow();
  });

  test("modifies file with correct CSS", async () => {
    const fs = require("fs");
    const callback = jest.fn();

    const properties = {
      "--border-radius": "4px",
    };

    const css = `.danger {
      border-radius: var(--border-radius, 5px);
    }`;

    const wrongProps = [
      {
        path: undefined,
        line: 2,
        key: "--border-radius",
        current: "5px",
        expected: "4px",
      },
    ];

    await run(css, { properties, write: true, callback });

    expect(fs.writeFile).toHaveBeenCalled();
    expect(callback).toHaveBeenCalledWith("foo", wrongProps);
  });
});


index.js
const { writeFile } = require("fs");

const { getCustomProp, noop } = require("./helper");

module.exports = validateCustomProps;
module.exports.postcss = true;

/**
 * Object to define an incorrect CSS custom property
 *
 * @typedef {Object} WrongProp
 * @property {string} path - absolute path of the CSS file
 * @property {number} line - line number containing the custom property
 * @property {string} key - key of the custom property
 * @property {string} current - current value of the custom property
 * @property {string} expected - expected correct value of the custom property
 */

/**
 * Callback to handle the response of the plugin
 *
 * @callback requestCallback
 * @param {Object} error - an error object
 * @param {WrongProp[]} data - a list of incorrect custom properties
 * @returns
 */

/**
 * Postcss plugin to validate fallback value of CSS custom properties
 *
 * @param {Object} opts - plugin options
 * @param {Object} opts.properties - CSS custom properties to validate values from
 * @param {boolean} opts.write [write=false] - update the file with updated CSS
 * @param {requestCallback} opts.callback [callback=() => {}] - callback to handle the response
 * @returns Object
 */
function validateCustomProps({ properties, write = false, callback = noop }) {
  if (!properties) {
    throw new Error("No properties present in options");
  }

  return {
    postcssPlugin: "postcss-customprop-validate",
    prepare(result) {
      const wrongProps = [];

      return {
        Declaration(decl) {
          const customProp = getCustomProp(decl);

          // only process custom properties
          if (!customProp) return;

          const { key, value } = customProp;
          const expected = properties[key];

          // only process if the custom property is defined in `properties` plugin option
          if (!expected) return;

          // only process if the fallback value is incorrect
          if (value === expected) return;

          wrongProps.push({
            path: decl.source.input.file,
            line: decl.source.start.line,
            key,
            current: value,
            expected,
          });

          decl.replaceWith({
            prop: decl.prop,
            raws: decl.raws,
            value: `var(${key}, ${expected})`,
          });
        },
        OnceExit(root) {
          result.wrongProps = wrongProps;

          const resultant = root.toResult().css;
          const { file } = root.source.input;

          if (write && wrongProps.length) {
            writeFile(file, resultant, (err) => callback(err, wrongProps));
          } else {
            callback(null, wrongProps);
          }
        },
      };
    },
  };
}


helper.js
const { toPlainObject, parse } = require("css-tree");

/**
 * Checks if a node is of type `Function`
 *
 * @param {Object} node node of AST
 * @param {string} node.type
 * @returns {boolean}
 */
const isFunction = ({ type }) => type === "Function";

/**
 * Checks if a node is of type `Identifier`
 *
 * @param {Object} node node of AST
 * @param {string} node.type
 * @returns {boolean}
 */
const isIdentifier = ({ type }) => type === "Identifier";

/**
 * Checks if a node is of type `Raw`
 *
 * @param {Object} node node of AST
 * @param {string} node.type
 * @returns {boolean}
 */
const isRaw = ({ type }) => type === "Raw";

/**
 * Removes newlines and extra spaces from a string
 *
 * @param {string} str
 * @returns {string}
 */
const getSingleLineValue = (str) =>
  str
    .replace(/(\r|\n)?\n/g, "")
    .replace(/\s{2,}/g, " ")
    .trim();

/**
 * Checks if the declaration's value contains a CSS custom property
 *
 * the custom properties are available as children of the 1st element of `tree` argument
 * and type of the object is `Function`
 *
 * @param {Object[]} tree AST transformed into a plain object
 * @returns {boolean}
 */
const hasCustomProp = (tree) => tree[0].children && !!tree.find(isFunction);

const noop = () => {};

/**
 * @typedef CustomProp
 * @property {string} key - key of the custom property
 * @property {string} value - value of the custom property
 */
/**
 * Validates if a Declaration node has a CSS custom property
 * If yes, returns its key and value
 * Else, returns `undefined`
 *
 * @param {Object} declaration declaration node
 * @returns {(CustomProp|undefined)}
 */
const getCustomProp = (declaration) => {
  const astObject = toPlainObject(
    parse(declaration.value, { context: "value" })
  ).children;

  if (!hasCustomProp(astObject)) return;

  const { children } = astObject[0];
  const identifier = children.find(isIdentifier);

  if (!identifier) return;

  const { name } = identifier;
  const raw = children.find(isRaw);
  const value = raw && getSingleLineValue(raw.value);

  return { key: name, value };
};

module.exports = {
  noop,
  getCustomProp,
};


values.js
const css = `.danger {
  --danger-radius: var(--border-radius, 4px);
  border-radius: var(--danger-radius);
}
.success {
  border: var(--custom-border, var(--border-radius, 5px) solid var(--color-red, #FA0000));
}
.info {
  color: var(--color-red, red);
}`;

const cssWithoutCustomProps = `.danger {
  color: hsla(292, 56%, 33%, 1);
  font-size: 1.2rem;
}`;

const cssWithUndefinedCustomProp = `.success {
  color: var(--color-green, #00FFAA);
  border-radius: var(--border-radius, 4px)
}`;

const properties = {
  "--border-radius": "4px",
  "--color-red": "#FF0000",
  "--custom-border":
    "var(--border-radius, 4px) solid var(--color-red, #FF0000)",
};

const wrongProps = [
  {
    path: undefined,
    line: 6,
    key: "--custom-border",
    current: "var(--border-radius, 5px) solid var(--color-red, #FA0000)",
    expected: "var(--border-radius, 4px) solid var(--color-red, #FF0000)",
  },
  {
    path: undefined,
    line: 9,
    key: "--color-red",
    current: "red",
    expected: "#FF0000",
  },
];

const resultantCSS = `.danger {
  --danger-radius: var(--border-radius, 4px);
  border-radius: var(--danger-radius);
}
.success {
  border: var(--custom-border, var(--border-radius, 4px) solid var(--color-red, #FF0000));
}
.info {
  color: var(--color-red, #FF0000);
}`;

module.exports = {
  css,
  cssWithUndefinedCustomProp,
  cssWithoutCustomProps,
  resultantCSS,
  wrongProps,
  properties,
};
